### ISSUE TITLE
If gi can't be imported, gtk2 reactor still fails (and vice versa)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5494]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5494|
|Type|defect|
|Created|2012-02-20 03:21:34Z|
|Branch|https://github.com/twisted/twisted/tree/gtk2-vs-gtk3-5494|

[#4558](https://github.com/twisted/twisted/issues/4558) made it impossible to import `twisted.internet.gtk2reactor` if `sys.modules` contains a "gi" key. This seems to overlook the fact that even a *failed* import of the `gi` module adds such a key to that dictionary.  Since the test suite tries to test `gireactor`, there will always be such a key, so `gtk2reactor` can never be imported by the test suite. The same issue would operate, only in reverse, if `gtk2reactor` was imported first.

<details><summary>Searchable metadata</summary>

```
trac-id__5494 5494
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__Twisted_12_1 Twisted-12.1
branch__branches_gtk2_vs_gtk3_5494 branches/gtk2-vs-gtk3-5494
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1329708094000000 1329708094000000
changetime__1331650023000000 1331650023000000
version__None None
owner__itamar itamar
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [33588]) Branching to 'gtk2-vs-gtk3-5494'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst removed owner|
|-|-|

Ready for review:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gtk2-vs-gtk3-5494

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @itamarst|
|-|-|

There's an effort to make a pygtk2 compatible wrapper around gi, and the current checks break that:
http://bazaar.launchpad.net/~stoq-dev/stoq/master/view/head:/stoq/lib/gicompat.py

So, we should make sure we support that too in this branch. Removing from review to ensure that.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst removed owner|
|-|-|

JP wants this in ASAP so we can have gtk2 tests running again, I'll do the compat layer fix separately.

**Comment by twisted-trac** [comment]
|<img alt="jdahlin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|jdahlin commented|
|-|-|

The patch looks good, but I'd switch the order of the parameters preventImports/errorMessage and always pass in the former by keyword, True/False are not really useful at the callsites.

I'd like a way to disable these checks, for gireactor/gtk3reactor, but that's probably another bug as discussed on irc.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @itamarst|
|-|-|

Sort of an improvement.  However, gtk2reactor tests still run on no build slaves (in other words, the issue has not been resolved).

The ones which might possibly run it seem to have a too-old version of `gi`?

First they skip the gtk3 tests because `'gi.repository.GLib' object has no attribute 'threads_init'` then they skip the gtk2 tests because `No module named gobject` (presumably due to the early behavior of `twisted/internet/gireactor.py`.

I feel as though our buildbot configuration is not doing us any favors here.  Do we need an alternate behavior for skipping or skip reporting?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

We are going to have to either setup a specific reactor buildstep for glib2 and gtk2, or setup a specific reactor buildstep for gi+gtk3. Technically we have the former but it has issues because it's too old to run gtk2 correctly! 

I reordered things to do the latter; since we don't have a sufficiently new buildslave to use gi anyway, may as well have gtk2 tests running. We will need to setup a buildslave for gi once we have one on a sufficiently modern OS. I will open a ticket for that and jdahlin's pygtk2 compat mode support when this ticket is merged.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst removed owner|
|-|-|

I also addressed jdahlin's review comments. Build run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gtk2-vs-gtk3-5494

gtk2 tests are running, e.g. here http://buildbot.twistedmatrix.com/builders/lucid64-py2.6-select/builds/1474

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess set owner to @itamarst|
|-|-|

Thanks for working on this, itamar. The code changes look good to me. A couple of mostly cosmetic comments:

 * `preventImports` defaults to a tuple but is described as a list in the docstring and is passed a list in the tests:
```
+def ensureNotImported(moduleNames, errorMessage, preventImports=()):
```


 * Can you add `@type` markup to the parameters in `ensureNotImported`?
 * missage => message in:
```
+    @raises: C{ImportError} with given error missage if a given module name
```

 * None could get epytext markup in:
```
+        If the specified modules have been set to None in C{sys.modules},
```

 * raise => raises in:
```
+        L{ensureNotImported} raise an exception.
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set status to `closed`|
|-|-|

(In [33732]) Merge gtk2-vs-gtk3-5494: Make sure gtk2 tests run.

Author: itamar
Review: exarkun, jesstess
Fixes: [#5494](https://github.com/twisted/twisted/issues/5494)

If gtk3 was missing, gtk2 tests still wouldn't run. This makes sure they will run. To run gtk3 or gi tests you will need to explicitly choose the applicable reactor for trial.
